filecoin-proofs/src/api/update.rs

```diff
@@ -652,3 +664,264 @@ pub fn verify_empty_sector_update_proof<Tree: 'static + MerkleTreeTrait<Hasher =
+        proofs.len(),
+    );
+
+    // Note that the proofs count here is not the same as the input
+    // proofs since the multi proof type takes the partitions into
+    // account
+    let target_proofs_len = get_aggregate_target_len(proofs.len());
+    ensure!(
+        target_proofs_len > 1,
+        "cannot aggregate less than two proofs"
+    );
+    trace!(
+        "aggregate_sector_update_proofs will pad proofs to target_len {}",
+        target_proofs_len
+    );
+
+    // If we're not at the pow2 target, duplicate the last proof until we are.
+    pad_proofs_to_target(&mut proofs, target_proofs_len)?;
+
+    let hashed_commitments = get_hashed_commitments(sector_update_inputs);
+    let srs_prover_key = get_stacked_srs_key::<Tree>(porep_config, proofs.len())?;
+    let aggregate_proof = EmptySectorUpdateCompound::<Tree>::aggregate_proofs(
+        &srs_prover_key,
+        &hashed_commitments,
+        proofs.as_slice(),
+        aggregate_version,
+    )?;
+    let mut aggregate_proof_bytes = Vec::new();
+    aggregate_proof.write(&mut aggregate_proof_bytes)?;
+
+    info!("aggregate_empty_sector_update_proofs:finish");
+
+    Ok(aggregate_proof_bytes)
+}
+
+pub fn verify_aggregate_sector_update_proofs<
+    Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>,
+>(
+    porep_config: &PoRepConfig,
+    aggregate_proof_bytes: AggregateSnarkProof,
+    inputs: &[SectorUpdateProofInputs],
+    sector_update_inputs: Vec<Vec<Fr>>,
+    aggregate_version: groth16::aggregate::AggregateVersion,
+) -> Result<bool> {
+    info!("verify_aggregate_sector_update_proofs:start");
+
+    info!(
+        "verify_aggregate_sector_update_proofs using API Version {}",
+        porep_config.api_version
+    );
+    ensure!(
+        porep_config.api_version >= ApiVersion::V1_2_0,
+        "Empty Sector Update proof aggregation is supported in ApiVersion 1.2.0 or later"
+    );
+    ensure!(
+        aggregate_version == groth16::aggregate::AggregateVersion::V2,
+        "Empty sector update aggregate verification requires SnarkPackV2"
+    );
+
+    let aggregate_proof =
+        groth16::aggregate::AggregateProof::read(std::io::Cursor::new(&aggregate_proof_bytes))?;
+
+    let aggregated_proofs_len = aggregate_proof.tmipp.gipa.nproofs as usize;
+
+    ensure!(aggregated_proofs_len != 0, "cannot verify zero proofs");
+    ensure!(
+        !sector_update_inputs.is_empty(),
+        "cannot verify with empty inputs"
+    );
+
+    trace!(
+        "verify_aggregate_sector_update_proofs called with len {}",
+        aggregated_proofs_len,
+    );
+
+    ensure!(
+        aggregated_proofs_len > 1,
+        "cannot verify less than two proofs"
+    );
+    ensure!(
+        aggregated_proofs_len == aggregated_proofs_len.next_power_of_two(),
+        "cannot verify non-pow2 aggregate seal proofs"
+    );
+
+    let num_inputs = sector_update_inputs.len();
+    let num_inputs_per_proof = get_aggregate_target_len(num_inputs) / aggregated_proofs_len;
```

Comment:
Just noting, I think `num_inputs_per_proof` should always be exactly `1`, i.e. each vector in `sector_update_inputs` are the public inputs to one Groth16 proof.